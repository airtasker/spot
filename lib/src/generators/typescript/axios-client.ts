import * as ts from "typescript";
import assertNever from "../../assert-never";
import {
  Api,
  DynamicPathComponent,
  Endpoint,
  objectType,
  PathComponent,
  stringConstant,
  Type,
  unionType
} from "../../models";
import { isVoid } from "../../validator";
import { outputTypeScriptSource } from "./ts-writer";
import { promiseTypeNode, typeNode } from "./types";
import {
  endpointPropertyTypeName,
  validateStatement,
  validatorName
} from "./validators";
import compact = require("lodash/compact");

const IMPORTED_AXIOS_NAME = "axios";

export function generateAxiosClientSource(api: Api): string {
  const typeNames = Object.keys(api.types);
  return outputTypeScriptSource([
    ts.createImportDeclaration(
      /*decorators*/ undefined,
      /*modifiers*/ undefined,
      ts.createImportClause(
        ts.createIdentifier(IMPORTED_AXIOS_NAME),
        /*namedBindings*/ undefined
      ),
      ts.createStringLiteral("axios")
    ),
    ...(typeNames.length > 0
      ? [
          ts.createImportDeclaration(
            /*decorators*/ undefined,
            /*modifiers*/ undefined,
            ts.createImportClause(
              /*name*/ undefined,
              ts.createNamedImports(
                typeNames.map(typeName =>
                  ts.createImportSpecifier(
                    /*propertyName*/ undefined,
                    ts.createIdentifier(typeName)
                  )
                )
              )
            ),
            ts.createStringLiteral("./types")
          )
        ]
      : []),
    ts.createImportDeclaration(
      /*decorators*/ undefined,
      /*modifiers*/ undefined,
      ts.createImportClause(
        /*name*/ undefined,
        ts.createNamespaceImport(ts.createIdentifier("validators"))
      ),
      ts.createStringLiteral("./validators")
    ),
    ...Object.entries(api.endpoints).map(([endpointName, endpoint]) =>
      generateEndpointFunction(api, endpointName, endpoint)
    )
  ]);
}

const REQUEST_PARAMETER = "request";

function generateEndpointFunction(
  api: Api,
  endpointName: string,
  endpoint: Endpoint
): ts.FunctionDeclaration {
  const includeRequest = !isVoid(api, endpoint.requestType);
  return ts.createFunctionDeclaration(
    /*decorators*/ undefined,
    [
      ts.createToken(ts.SyntaxKind.ExportKeyword),
      ts.createToken(ts.SyntaxKind.AsyncKeyword)
    ],
    /*asteriskToken*/ undefined,
    endpointName,
    /*typeParameters*/ undefined,
    [
      ...(includeRequest
        ? [
            ts.createParameter(
              /*decorators*/ undefined,
              /*modifiers*/ undefined,
              /*dotDotDotToken*/ undefined,
              REQUEST_PARAMETER,
              /*questionToken*/ undefined,
              typeNode(endpoint.requestType)
            )
          ]
        : []),
      ...compact(
        endpoint.path.map(
          pathComponent =>
            pathComponent.kind === "dynamic"
              ? ts.createParameter(
                  /*decorators*/ undefined,
                  /*modifiers*/ undefined,
                  /*dotDotDotToken*/ undefined,
                  pathComponent.name,
                  /*questionToken*/ undefined,
                  typeNode(pathComponent.type)
                )
              : null
        )
      ),
      ...Object.entries(endpoint.headers).map(([headerName, header]) =>
        ts.createParameter(
          /*decorators*/ undefined,
          /*modifiers*/ undefined,
          /*dotDotDotToken*/ undefined,
          headerName,
          /*questionToken*/ undefined,
          typeNode(header.type)
        )
      ),
      ...endpoint.queryParams.map(queryParam =>
        ts.createParameter(
          /*decorators*/ undefined,
          /*modifiers*/ undefined,
          /*dotDotDotToken*/ undefined,
          queryParam.name,
          /*questionToken*/ undefined,
          typeNode(queryParam.type)
        )
      )
    ],
    promiseTypeNode(unionType(...generateReturnTypes(endpoint))),
    generateEndpointBody(endpointName, endpoint, includeRequest)
  );
}

function generateReturnTypes(endpoint: Endpoint): Type[] {
  return [
    objectType({
      kind: stringConstant("success"),
      data: endpoint.responseType
    }),
    objectType({
      kind: stringConstant("unknown-error"),
      data: endpoint.genericErrorType
    }),
    ...Object.entries(endpoint.specificErrorTypes).map(
      ([name, specificError]) =>
        objectType({
          kind: stringConstant(name),
          data: specificError.type
        })
    )
  ];
}

function generateEndpointBody(
  endpointName: string,
  endpoint: Endpoint,
  includeRequest: boolean
): ts.FunctionBody {
  return ts.createBlock(
    [
      ...(includeRequest
        ? [generateRequestValidation(endpointName, endpoint)]
        : []),
      ...compact(
        endpoint.path.map(
          pathComponent =>
            pathComponent.kind === "dynamic"
              ? generatePathParameterValidation(endpointName, pathComponent)
              : null
        )
      ),
      ...Object.keys(endpoint.headers).map(headerName =>
        generateHeaderValidation(endpointName, headerName)
      ),
      generateAxiosCall(endpoint, includeRequest),
      generateSwitchStatus(endpointName, endpoint)
    ],
    /*multiLine*/ true
  );
}

function generateRequestValidation(
  endpointName: string,
  endpoint: Endpoint
): ts.Statement {
  return validateStatement(
    ts.createIdentifier(REQUEST_PARAMETER),
    validatorName(endpointPropertyTypeName(endpointName, REQUEST_PARAMETER)),
    "Invalid request"
  );
}

function generateHeaderValidation(
  endpointName: string,
  headerName: string
): ts.Statement {
  return validateStatement(
    ts.createIdentifier(headerName),
    validatorName(endpointPropertyTypeName(endpointName, "header", headerName)),
    `Invalid parameter ${headerName}`
  );
}

function generatePathParameterValidation(
  endpointName: string,
  pathComponent: DynamicPathComponent
): ts.Statement {
  return validateStatement(
    ts.createIdentifier(pathComponent.name),
    validatorName(
      endpointPropertyTypeName(endpointName, "param", pathComponent.name)
    ),
    `Invalid parameter ${pathComponent.name}`
  );
}

const RESPONSE_NAME = "response";
const RESPONSE_STATUS_CODE = ts.createPropertyAccess(
  ts.createIdentifier(RESPONSE_NAME),
  "status"
);
const RESPONSE_DATA = ts.createPropertyAccess(
  ts.createIdentifier(RESPONSE_NAME),
  "data"
);

function generateAxiosCall(
  endpoint: Endpoint,
  includeRequest: boolean
): ts.Statement {
  return ts.createVariableStatement(
    /*modifiers*/ undefined,
    ts.createVariableDeclarationList(
      [
        ts.createVariableDeclaration(
          RESPONSE_NAME,
          /*type*/ undefined,
          ts.createAwait(
            ts.createCall(
              ts.createIdentifier(IMPORTED_AXIOS_NAME),
              /*typeArguments*/ undefined,
              [
                ts.createObjectLiteral(
                  [
                    ts.createPropertyAssignment(
                      "url",
                      generatePath(endpoint.path)
                    ),
                    ts.createPropertyAssignment(
                      "method",
                      ts.createStringLiteral(endpoint.method)
                    ),
                    ts.createPropertyAssignment(
                      "responseType",
                      ts.createStringLiteral("json")
                    ),
                    ts.createPropertyAssignment(
                      "headers",
                      ts.createObjectLiteral(
                        Object.entries(endpoint.headers).map(
                          ([headerName, header]) =>
                            ts.createPropertyAssignment(
                              header.headerFieldName,
                              ts.createIdentifier(headerName)
                            )
                        )
                      )
                    ),
                    ...(includeRequest
                      ? [
                          ts.createPropertyAssignment(
                            "data",
                            ts.createIdentifier(REQUEST_PARAMETER)
                          )
                        ]
                      : []),
                    ts.createPropertyAssignment(
                      "validateStatus",
                      ts.createArrowFunction(
                        /*modifiers*/ undefined,
                        /*typeParameters*/ undefined,
                        [],
                        /*type*/ undefined,
                        /*equalsGreaterThanToken*/ undefined,
                        ts.createLiteral(true)
                      )
                    )
                  ],
                  /*multiLine*/ true
                )
              ]
            )
          )
        )
      ],
      ts.NodeFlags.Const
    )
  );
}

function generatePath(path: PathComponent[]): ts.Expression {
  let result: ts.Expression | null = null;
  for (const pathComponent of path) {
    let append;
    switch (pathComponent.kind) {
      case "static":
        append = ts.createStringLiteral(pathComponent.content);
        break;
      case "dynamic":
        append = ts.createIdentifier(pathComponent.name);
        break;
      default:
        throw assertNever(pathComponent);
    }
    if (!result) {
      result = append;
    } else {
      result = ts.createBinary(result, ts.SyntaxKind.PlusToken, append);
    }
  }
  if (!result) {
    result = ts.createStringLiteral("");
  }
  return result;
}

function generateSwitchStatus(
  endpointName: string,
  endpoint: Endpoint
): ts.Statement {
  return ts.createSwitch(
    RESPONSE_STATUS_CODE,
    ts.createCaseBlock([
      ...Object.entries(endpoint.specificErrorTypes).map(
        ([name, specificError]) =>
          ts.createCaseClause(
            ts.createNumericLiteral(specificError.statusCode.toString(10)),
            [
              validateStatement(
                RESPONSE_DATA,
                validatorName(
                  endpointPropertyTypeName(endpointName, "specificError", name)
                ),
                `Invalid response for status code ${specificError.statusCode}`
              ),
              ts.createReturn(
                ts.createObjectLiteral(
                  [
                    ts.createPropertyAssignment(
                      "kind",
                      ts.createStringLiteral(name)
                    ),
                    ts.createPropertyAssignment("data", RESPONSE_DATA)
                  ],
                  /*multiLine*/ true
                )
              )
            ]
          )
      ),
      ts.createDefaultClause([
        ts.createIf(
          ts.createLogicalAnd(
            ts.createBinary(
              RESPONSE_STATUS_CODE,
              ts.SyntaxKind.GreaterThanEqualsToken,
              ts.createNumericLiteral("200")
            ),
            ts.createBinary(
              RESPONSE_STATUS_CODE,
              ts.SyntaxKind.LessThanToken,
              ts.createNumericLiteral("300")
            )
          ),
          ts.createBlock(
            [
              validateStatement(
                RESPONSE_DATA,
                validatorName(
                  endpointPropertyTypeName(endpointName, "response")
                ),
                `Invalid response for successful status code`
              ),
              ts.createReturn(
                ts.createObjectLiteral(
                  [
                    ts.createPropertyAssignment(
                      "kind",
                      ts.createStringLiteral("success")
                    ),
                    ts.createPropertyAssignment("data", RESPONSE_DATA)
                  ],
                  /*multiLine*/ true
                )
              )
            ],
            /*multiLine*/ true
          ),
          ts.createBlock(
            [
              validateStatement(
                RESPONSE_DATA,
                validatorName(
                  endpointPropertyTypeName(endpointName, "genericError")
                ),
                `Invalid response for unknown error`
              ),
              ts.createReturn(
                ts.createObjectLiteral(
                  [
                    ts.createPropertyAssignment(
                      "kind",
                      ts.createStringLiteral("unknown-error")
                    ),
                    ts.createPropertyAssignment("data", RESPONSE_DATA)
                  ],
                  /*multiLine*/ true
                )
              )
            ],
            /*multiLine*/ true
          )
        )
      ])
    ])
  );
}
