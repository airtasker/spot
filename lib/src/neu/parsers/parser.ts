import * as path from "path";
import { CompilerOptions, Project, SourceFile, ts } from "ts-morph";
import { ApiConfig } from "../../syntax/api";
import { Contract, Endpoint, Security } from "../definitions";
import { LociTable } from "../locations";
import { TypeTable } from "../types";
import { parseEndpoint } from "./endpoint-parser";
import {
  getClassWithDecoratorOrThrow,
  getDecoratorConfigOrThrow,
  getJsDoc,
  getObjLiteralPropOrThrow,
  getPropertyWithDecorator,
  getPropValueAsStringOrThrow,
  getSelfAndLocalDependencies
} from "./parser-helpers";
import { parseSecurityHeader } from "./security-header-parser";

export function parse(sourcePath: string): Contract {
  const project = createProject();

  // Add all dependent files that the project requires
  const sourceFile = project.addExistingSourceFile(sourcePath);
  project.resolveSourceFileDependencies();

  // Validate that the project has no TypeScript syntax errors
  validateProject(project);

  return parseRootSourceFile(sourceFile);
}

/**
 * Parse a root source file to return a contract.
 */
function parseRootSourceFile(file: SourceFile): Contract {
  const lociTable = new LociTable();
  const typeTable = new TypeTable();

  const klass = getClassWithDecoratorOrThrow(file, "api");
  const decorator = klass.getDecoratorOrThrow("api");
  const decoratorConfig = getDecoratorConfigOrThrow(decorator);
  const nameProp = getObjLiteralPropOrThrow<ApiConfig>(decoratorConfig, "name");
  const nameLiteral = getPropValueAsStringOrThrow(nameProp);
  const descriptionDoc = getJsDoc(klass);

  const securityHeaderProp = getPropertyWithDecorator(klass, "securityHeader");

  const security: Security = {
    header:
      securityHeaderProp &&
      parseSecurityHeader(securityHeaderProp, typeTable, lociTable)
  };

  // Add location data
  lociTable.addMorphNode(LociTable.apiClassKey(), klass);
  lociTable.addMorphNode(LociTable.apiDecoratorKey(), decorator);
  lociTable.addMorphNode(LociTable.apiNameKey(), nameProp);
  if (descriptionDoc) {
    lociTable.addMorphNode(LociTable.apiDescriptionKey(), descriptionDoc);
  }

  // Resolve all related files
  const projectFiles = getSelfAndLocalDependencies(file);

  // Parse all endpoints
  const endpoints = projectFiles
    .reduce<Endpoint[]>(
      (acc, currentFile) =>
        acc.concat(
          currentFile
            .getClasses()
            .filter(k => k.getDecorator("endpoint") !== undefined)
            .map(k => parseEndpoint(k, typeTable, lociTable))
        ),
      []
    )
    .sort((a, b) => {
      if (a.name === b.name) {
        throw new Error(`Duplicate endpoint detected: ${a.name}`);
      }
      return b.name > a.name ? -1 : 1;
    });

  return {
    name: nameLiteral.getLiteralText(),
    description: descriptionDoc && descriptionDoc.getComment(),
    types: typeTable.toArray(),
    security,
    endpoints
  };
}

/**
 * Create a new project configured for Spot
 */
function createProject(): Project {
  const compilerOptions: CompilerOptions = {
    target: ts.ScriptTarget.ESNext,
    module: ts.ModuleKind.CommonJS,
    strict: true,
    noImplicitAny: true,
    strictNullChecks: true,
    strictFunctionTypes: true,
    strictPropertyInitialization: true,
    noImplicitThis: true,
    alwaysStrict: true,
    noImplicitReturns: true,
    noFallthroughCasesInSwitch: true,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    experimentalDecorators: true,
    baseUrl: "./",
    paths: {
      "@airtasker/spot": [path.join(__dirname, "../lib")]
    }
  };

  // Creates a new typescript program in memory
  return new Project({ compilerOptions });
}

/**
 * Validate an AST project's correctness.
 *
 * @param project an AST project
 */
function validateProject(project: Project) {
  const diagnostics = project.getPreEmitDiagnostics();
  if (diagnostics.length > 0) {
    throw new Error(
      diagnostics
        .map(diagnostic => {
          const message = diagnostic.getMessageText();
          return typeof message === "string"
            ? message
            : message.getMessageText();
        })
        .join("\n")
    );
  }
}
