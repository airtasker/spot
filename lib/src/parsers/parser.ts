import path from "path";
import { CompilerOptions, Project, SourceFile, ts } from "ts-morph";
import { Locatable } from "../models/locatable";
import { ContractNode, EndpointNode } from "../models/nodes";
import { parseApi } from "./nodes/api-parser";
import { parseEndpoint } from "./nodes/endpoint-parser";
import { extractJsDocComment } from "./utilities/parser-utility";
import {
  parseInterfaceDeclaration,
  parseTypeNode
} from "./utilities/type-parser";
import {
  retrieveTypeReferencesFromEndpoints,
  uniqueReferences
} from "./utilities/type-reference-resolver";

export function parse(sourcePath: string): ContractNode {
  const compilerOptions: CompilerOptions = {
    target: ts.ScriptTarget.ESNext,
    module: ts.ModuleKind.CommonJS,
    strict: true,
    noImplicitAny: true,
    strictNullChecks: true,
    strictFunctionTypes: true,
    strictPropertyInitialization: true,
    noImplicitThis: true,
    alwaysStrict: true,
    noImplicitReturns: true,
    noFallthroughCasesInSwitch: true,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    experimentalDecorators: true,
    baseUrl: "./",
    paths: {
      "@airtasker/spot": [path.join(__dirname, "../lib")]
    }
  };

  // Creates a new typescript program in memory
  const project = new Project({ compilerOptions });
  const sourceFile = project.addExistingSourceFile(sourcePath);

  // Add all dependent files that the project requires
  project.resolveSourceFileDependencies();

  // Validate that the project has no TypeScript syntax errors
  validateProject(project);

  return parseRootSourceFile(sourceFile, project);
}

/**
 * Parse a root source file to return a contract.
 */
function parseRootSourceFile(
  file: SourceFile,
  projectContext: Project
): ContractNode {
  const apiClasses = file
    .getClasses()
    .filter(klass => klass.getDecorator("api") !== undefined);

  if (apiClasses.length !== 1) {
    throw new Error(
      `expected class decorated with @api to be defined exactly once in root source file, found ${
        apiClasses.length
      } usages`
    );
  }
  const api = parseApi(apiClasses[0]);

  const endpoints = parseRecursively(file);

  const referenceTypes = uniqueReferences(
    retrieveTypeReferencesFromEndpoints(endpoints, projectContext)
  );

  // Construct the equivalent TypeNode for each reference type
  const types = referenceTypes.map(referenceType => {
    const file = projectContext.getSourceFileOrThrow(referenceType.location);

    const typeAlias = file.getTypeAlias(referenceType.name);
    if (typeAlias !== undefined) {
      const name = typeAlias.getName();
      const description = extractJsDocComment(typeAlias);
      const type = parseTypeNode(typeAlias.getTypeNodeOrThrow());
      return { description, name, type };
    }

    const interfaceDeclaration = file.getInterface(referenceType.name);
    if (interfaceDeclaration) {
      const name = interfaceDeclaration.getName();
      const description = extractJsDocComment(interfaceDeclaration);
      const type = parseInterfaceDeclaration(interfaceDeclaration);
      return { description, name, type };
    }

    throw new Error("unexpected type reference");
  });

  return { api, endpoints, types };
}

/**
 * Parses a file and its local imports/exports recursively, looking for endpoint definitions.
 *
 * @param file The current source file (starting with the root file where the API is defined).
 * @param visitedPaths The list of paths that were already visited, to avoid an infinite loop
 * with circular dependencies.
 */
function parseRecursively(
  file: SourceFile,
  visitedPaths: string[] = []
): Array<Locatable<EndpointNode>> {
  const filePath = file.getFilePath();
  if (visitedPaths.indexOf(filePath) !== -1) {
    // Don't visit the same file multiple times.
    return [];
  } else {
    visitedPaths.push(filePath);
  }

  const endpoints = file
    .getClasses()
    .filter(klass => klass.getDecorator("endpoint") !== undefined)
    .map(endpointClass => parseEndpoint(endpointClass));

  const importedFiles = file
    .getImportDeclarations()
    // We only care about local imports.
    .filter(i => i.getModuleSpecifierValue().startsWith("."))
    .map(myImport => myImport.getModuleSpecifierSourceFileOrThrow());

  const exportedFiles = file
    .getExportDeclarations()
    // We only care about local imports.
    .filter(e => {
      const moduleSpecifierValue = e.getModuleSpecifierValue();
      return moduleSpecifierValue && moduleSpecifierValue.startsWith(".");
    })
    .map(myImport => myImport.getModuleSpecifierSourceFileOrThrow());

  return [...importedFiles, ...exportedFiles].reduce<
    Array<Locatable<EndpointNode>>
  >(
    (endpointsAcc, currentFile) =>
      parseRecursively(currentFile, visitedPaths).concat(endpointsAcc),
    endpoints
  );
}

/**
 * Validate an AST project's correctness.
 *
 * @param project an AST project
 */
function validateProject(project: Project) {
  const diagnostics = project.getPreEmitDiagnostics();
  if (diagnostics.length > 0) {
    throw new Error(
      diagnostics
        .map(diagnostic => {
          const message = diagnostic.getMessageText();
          return typeof message === "string"
            ? message
            : message.getMessageText();
        })
        .join("\n")
    );
  }
}
