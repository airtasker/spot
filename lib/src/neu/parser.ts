import * as path from "path";
import { CompilerOptions, Project, ts } from "ts-morph";
import { Contract } from "./definitions";
import { parseContract } from "./parsers/contract-parser";

export function parse(sourcePath: string): Contract {
  const project = createProject();

  // Add all dependent files that the project requires
  const sourceFile = project.addExistingSourceFile(sourcePath);
  project.resolveSourceFileDependencies();

  // Validate that the project has no TypeScript syntax errors
  validateProject(project);

  const result = parseContract(sourceFile);

  if (result.isErr()) throw result.unwrapErr();

  return result.unwrap().contract;
}

/**
 * Create a new project configured for Spot
 */
function createProject(): Project {
  const compilerOptions: CompilerOptions = {
    target: ts.ScriptTarget.ESNext,
    module: ts.ModuleKind.CommonJS,
    strict: true,
    noImplicitAny: true,
    strictNullChecks: true,
    strictFunctionTypes: true,
    strictPropertyInitialization: true,
    noImplicitThis: true,
    alwaysStrict: true,
    noImplicitReturns: true,
    noFallthroughCasesInSwitch: true,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    experimentalDecorators: true,
    baseUrl: "./",
    paths: {
      "@airtasker/spot": [path.join(__dirname, "../lib")]
    }
  };

  // Creates a new typescript program in memory
  return new Project({ compilerOptions });
}

/**
 * Validate an AST project's correctness.
 *
 * @param project an AST project
 */
function validateProject(project: Project) {
  const diagnostics = project.getPreEmitDiagnostics();
  if (diagnostics.length > 0) {
    throw new Error(
      diagnostics
        .map(diagnostic => {
          const message = diagnostic.getMessageText();
          return typeof message === "string"
            ? message
            : message.getMessageText();
        })
        .join("\n")
    );
  }
}
