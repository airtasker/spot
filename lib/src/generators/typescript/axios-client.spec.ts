import * as fs from "fs-extra";
import * as path from "path";
import { parsePath } from "../../parser";
import { generateAxiosClientSource } from "./axios-client";

const EXAMPLES_DIR = path.join(__dirname, "..", "..", "..", "examples");

describe("TypeScript axios client generator", () => {
  describe("produces valid code", () => {
    for (const testCaseName of fs.readdirSync(EXAMPLES_DIR)) {
      test(testCaseName, async () => {
        const api = await parsePath(
          path.join(EXAMPLES_DIR, testCaseName, "api.ts")
        );
        const source = generateAxiosClientSource(api);
        expect(source).toMatchSnapshot();
      });
    }
  });
});
