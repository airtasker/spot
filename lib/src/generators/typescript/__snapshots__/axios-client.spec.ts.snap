// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScript axios client generator produces valid code multi-file 1`] = `
"import axios from \\"axios\\";

import { CreateUserRequest, CreateUserResponse } from \\"./types\\";

import * as validators from \\"./validators\\";

export async function createUser(request: CreateUserRequest, authToken: string | void): Promise<{
    kind: \\"success\\";
    data: CreateUserResponse;
} | {
    kind: \\"unknown-error\\";
    data?: void;
}> {
    if (!validators.validateCreateUser_request(request)) {
        throw new Error(\`Invalid request: \${JSON.stringify(request, null, 2)}\`);
    }
    if (!validators.validateCreateUser_headerAuthToken(authToken)) {
        throw new Error(\`Invalid parameter authToken: \${JSON.stringify(authToken, null, 2)}\`);
    }
    const response = await axios({
        url: \\"/users/create\\",
        method: \\"POST\\",
        responseType: \\"json\\",
        headers: { \\"Authorization\\": authToken },
        data: request,
        validateStatus: () => true
    });
    switch (response.status) {
        default: if (response.status >= 200 && response.status < 300) {
            if (!validators.validateCreateUser_response(response.data)) {
                throw new Error(\`Invalid response for successful status code: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"success\\",
                data: response.data
            };
        }
        else {
            if (!validators.validateCreateUser_genericError(response.data)) {
                throw new Error(\`Invalid response for unknown error: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"unknown-error\\",
                data: response.data
            };
        }
    }
}

export async function deleteUser(userId: number, authToken: string): Promise<{
    kind: \\"success\\";
    data: null;
} | {
    kind: \\"unknown-error\\";
    data: {
        message: string;
    };
} | {
    kind: \\"forbidden\\";
    data: {
        message: string;
        signedInAs: string;
    };
}> {
    if (!validators.validateDeleteUser_paramUserId(userId)) {
        throw new Error(\`Invalid parameter userId: \${JSON.stringify(userId, null, 2)}\`);
    }
    if (!validators.validateDeleteUser_headerAuthToken(authToken)) {
        throw new Error(\`Invalid parameter authToken: \${JSON.stringify(authToken, null, 2)}\`);
    }
    const response = await axios({
        url: \\"/users/\\" + userId + \\"-confirmed\\",
        method: \\"DELETE\\",
        responseType: \\"json\\",
        headers: { \\"Authorization\\": authToken },
        validateStatus: () => true
    });
    switch (response.status) {
        case 403:
            if (!validators.validateDeleteUser_specificErrorForbidden(response.data)) {
                throw new Error(\`Invalid response for status code 403: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"forbidden\\",
                data: response.data
            };
        default: if (response.status >= 200 && response.status < 300) {
            if (!validators.validateDeleteUser_response(response.data)) {
                throw new Error(\`Invalid response for successful status code: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"success\\",
                data: response.data
            };
        }
        else {
            if (!validators.validateDeleteUser_genericError(response.data)) {
                throw new Error(\`Invalid response for unknown error: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"unknown-error\\",
                data: response.data
            };
        }
    }
}

export async function findUsers(limit: number, search_term: string | void): Promise<{
    kind: \\"success\\";
    data: {
        name: string;
        age?: number;
    }[];
} | {
    kind: \\"unknown-error\\";
    data?: void;
}> {
    const response = await axios({
        url: \\"/users\\",
        method: \\"GET\\",
        responseType: \\"json\\",
        headers: {},
        validateStatus: () => true
    });
    switch (response.status) {
        default: if (response.status >= 200 && response.status < 300) {
            if (!validators.validateFindUsers_response(response.data)) {
                throw new Error(\`Invalid response for successful status code: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"success\\",
                data: response.data
            };
        }
        else {
            if (!validators.validateFindUsers_genericError(response.data)) {
                throw new Error(\`Invalid response for unknown error: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"unknown-error\\",
                data: response.data
            };
        }
    }
}

export async function getUser(userId: number): Promise<{
    kind: \\"success\\";
    data: {
        name: string;
        age?: number;
    };
} | {
    kind: \\"unknown-error\\";
    data?: void;
}> {
    if (!validators.validateGetUser_paramUserId(userId)) {
        throw new Error(\`Invalid parameter userId: \${JSON.stringify(userId, null, 2)}\`);
    }
    const response = await axios({
        url: \\"/users/\\" + userId,
        method: \\"GET\\",
        responseType: \\"json\\",
        headers: {},
        validateStatus: () => true
    });
    switch (response.status) {
        default: if (response.status >= 200 && response.status < 300) {
            if (!validators.validateGetUser_response(response.data)) {
                throw new Error(\`Invalid response for successful status code: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"success\\",
                data: response.data
            };
        }
        else {
            if (!validators.validateGetUser_genericError(response.data)) {
                throw new Error(\`Invalid response for unknown error: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"unknown-error\\",
                data: response.data
            };
        }
    }
}"
`;

exports[`TypeScript axios client generator produces valid code single-file 1`] = `
"import axios from \\"axios\\";

import { CreateUserRequest, CreateUserResponse } from \\"./types\\";

import * as validators from \\"./validators\\";

export async function createUser(request: CreateUserRequest, authToken: string | void): Promise<{
    kind: \\"success\\";
    data: CreateUserResponse;
} | {
    kind: \\"unknown-error\\";
    data?: void;
}> {
    if (!validators.validateCreateUser_request(request)) {
        throw new Error(\`Invalid request: \${JSON.stringify(request, null, 2)}\`);
    }
    if (!validators.validateCreateUser_headerAuthToken(authToken)) {
        throw new Error(\`Invalid parameter authToken: \${JSON.stringify(authToken, null, 2)}\`);
    }
    const response = await axios({
        url: \\"/users/create\\",
        method: \\"POST\\",
        responseType: \\"json\\",
        headers: { \\"Authorization\\": authToken },
        data: request,
        validateStatus: () => true
    });
    switch (response.status) {
        default: if (response.status >= 200 && response.status < 300) {
            if (!validators.validateCreateUser_response(response.data)) {
                throw new Error(\`Invalid response for successful status code: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"success\\",
                data: response.data
            };
        }
        else {
            if (!validators.validateCreateUser_genericError(response.data)) {
                throw new Error(\`Invalid response for unknown error: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"unknown-error\\",
                data: response.data
            };
        }
    }
}

export async function findUsers(limit: number, search_term: string | void): Promise<{
    kind: \\"success\\";
    data: {
        name: string;
        age?: number;
    }[];
} | {
    kind: \\"unknown-error\\";
    data?: void;
}> {
    const response = await axios({
        url: \\"/users\\",
        method: \\"GET\\",
        responseType: \\"json\\",
        headers: {},
        validateStatus: () => true
    });
    switch (response.status) {
        default: if (response.status >= 200 && response.status < 300) {
            if (!validators.validateFindUsers_response(response.data)) {
                throw new Error(\`Invalid response for successful status code: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"success\\",
                data: response.data
            };
        }
        else {
            if (!validators.validateFindUsers_genericError(response.data)) {
                throw new Error(\`Invalid response for unknown error: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"unknown-error\\",
                data: response.data
            };
        }
    }
}

export async function getUser(userId: number): Promise<{
    kind: \\"success\\";
    data: {
        name: string;
        age?: number;
    };
} | {
    kind: \\"unknown-error\\";
    data?: void;
}> {
    if (!validators.validateGetUser_paramUserId(userId)) {
        throw new Error(\`Invalid parameter userId: \${JSON.stringify(userId, null, 2)}\`);
    }
    const response = await axios({
        url: \\"/users/\\" + userId,
        method: \\"GET\\",
        responseType: \\"json\\",
        headers: {},
        validateStatus: () => true
    });
    switch (response.status) {
        default: if (response.status >= 200 && response.status < 300) {
            if (!validators.validateGetUser_response(response.data)) {
                throw new Error(\`Invalid response for successful status code: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"success\\",
                data: response.data
            };
        }
        else {
            if (!validators.validateGetUser_genericError(response.data)) {
                throw new Error(\`Invalid response for unknown error: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"unknown-error\\",
                data: response.data
            };
        }
    }
}

export async function deleteUser(userId: number, authToken: string): Promise<{
    kind: \\"success\\";
    data: null;
} | {
    kind: \\"unknown-error\\";
    data: {
        message: string;
    };
} | {
    kind: \\"forbidden\\";
    data: {
        message: string;
        signedInAs: string;
    };
}> {
    if (!validators.validateDeleteUser_paramUserId(userId)) {
        throw new Error(\`Invalid parameter userId: \${JSON.stringify(userId, null, 2)}\`);
    }
    if (!validators.validateDeleteUser_headerAuthToken(authToken)) {
        throw new Error(\`Invalid parameter authToken: \${JSON.stringify(authToken, null, 2)}\`);
    }
    const response = await axios({
        url: \\"/users/\\" + userId + \\"-confirmed\\",
        method: \\"DELETE\\",
        responseType: \\"json\\",
        headers: { \\"Authorization\\": authToken },
        validateStatus: () => true
    });
    switch (response.status) {
        case 403:
            if (!validators.validateDeleteUser_specificErrorForbidden(response.data)) {
                throw new Error(\`Invalid response for status code 403: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"forbidden\\",
                data: response.data
            };
        default: if (response.status >= 200 && response.status < 300) {
            if (!validators.validateDeleteUser_response(response.data)) {
                throw new Error(\`Invalid response for successful status code: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"success\\",
                data: response.data
            };
        }
        else {
            if (!validators.validateDeleteUser_genericError(response.data)) {
                throw new Error(\`Invalid response for unknown error: \${JSON.stringify(response.data, null, 2)}\`);
            }
            return {
                kind: \\"unknown-error\\",
                data: response.data
            };
        }
    }
}"
`;
